pkg/reconciler/taskrun/resources/taskref.go

```diff
@@ -17,13 +17,78 @@ limitations under the License.
 package resources
 
 import (
+	"context"
 	"fmt"
+	"strings"
 
+	"github.com/google/go-containerregistry/pkg/authn/k8schain"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
 	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
 	clientset "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	"github.com/tektoncd/pipeline/pkg/remote/oci"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
 )
 
+// GetTaskFunc is a factory function that will use the given TaskRef as context to return a valid GetTask function. It
+// also requires a kubeclient, tektonclient, namespace, and service account in case it needs to find that task in
+// cluster or authorize against an external repositroy. It will figure out whether it needs to look in the cluster or in
+// a remote image to fetch the  reference. It will also return the "kind" of the task being referenced.
+func GetTaskFunc(k8s kubernetes.Interface, tekton clientset.Interface, tr *v1beta1.TaskRef, namespace, saName string) (GetTask, v1beta1.TaskKind, error) {
+	kind := v1alpha1.NamespacedTaskKind
+	if tr != nil && tr.Kind != "" {
+		kind = tr.Kind
+	}
+
+	switch {
+	case tr != nil && tr.Bundle != "":
+		// If there is a bundle url at all, construct an OCI resolver to fetch the task.
+		kc, err := k8schain.New(k8s, k8schain.Options{
+			Namespace:          namespace,
+			ServiceAccountName: saName,
+		})
+		if err != nil {
+			return nil, kind, fmt.Errorf("failed to get keychain: %w", err)
+		}
+		resolver := oci.NewResolver(tr.Bundle, kc)
+		// Return an inline function that implements GetTask by calling Resolver.Get with the specified task type and
+		// casting it to a TaskInterface.
+		return func(name string) (v1beta1.TaskInterface, error) {
+			obj, err := resolver.Get(strings.ToLower(string(kind)), name)
+			if err != nil {
+				return nil, err
+			}
+
+			// We need to coerce this object into a v1beta1 interface.
+			if ti, ok := obj.(v1beta1.TaskInterface); ok {
+				return ti, nil
+			}
+
+			if task, ok := obj.(*v1alpha1.Task); ok {
+				betaTask := &v1beta1.Task{}
+				err := task.ConvertTo(context.Background(), betaTask)
+				return betaTask, err
+			}
```

Comment:
Why are we converting alpha task to beta here? Should we check the API Version if specified in `taskRef`? Also, could there be conflict in API version, one in `taskRef` could be different from the remote spec?